Workflow/Sources/WorkflowObserver.swift

```diff
@@ -0,0 +1,262 @@
+ */
+
+import Foundation
+
+// MARK: - WorkflowObserver
+
+/// The `WorkflowObserver` protocol provides an interface to receive updates during the runtime's
+/// execution loop. All requirements are optional, and no-ops are provided by default.
+public protocol WorkflowObserver {
+    /// Indicates the start of a `WorkflowSession`, which tracks the life of the underlying `WorkflowNode` used to provide renderings for a given `Workflow` type and rendering key.
+    /// - Parameter session: The `WorkflowSession` that began.
+    func sessionDidBegin(
+        _ session: WorkflowSession
+    )
+
+    /// Marks the end of a `WorkflowSession`, indicating that the corresponding `WorkflowNode` has been removed from the tree of Workflows.
+    /// - Parameter session: The `WorkflowSession` that ended.
+    func sessionDidEnd(
+        _ session: WorkflowSession
+    )
+
+    /// Indicates a `Workflow` produced its initial state value.
+    /// - Parameters:
+    ///   - workflow: The `Workflow` that just produced its initial state.
+    ///   - initialState: The `State` that was created.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`
+    func workflowDidMakeInitialState<WorkflowType: Workflow>(
+        _ workflow: WorkflowType,
+        initialState: WorkflowType.State,
+        session: WorkflowSession
+    )
+
+    /// Called before a `Workflow` is queried for its latest `Rendering`.
+    /// - Parameters:
+    ///   - workflow: The `Workflow` that is about to be render.
+    ///   - state: The corresponding `State` that will be used during the render call.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    /// - Returns: An optional closure to be called immediately after the new `Rendering` is produced, which takes the rendering as the only parameter.
+    func workflowWillRender<WorkflowType: Workflow>(
+        _ workflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    ) -> ((WorkflowType.Rendering) -> Void)?
+
+    /// Called after an existing `Workflow` is updated.
+    /// - Parameters:
+    ///   - oldWorkflow: The previous `Workflow`
+    ///   - newWorkflow: The new `Workflow`
+    ///   - state: The state **after** the update has occurred.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    func workflowDidChange<WorkflowType: Workflow>(
+        from oldWorkflow: WorkflowType,
+        to newWorkflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    )
+
+    /// Called after a `WorkflowAction` is received, but **before** it has been handled and propagated up the tree.
+    /// - Parameters:
+    ///   - action: The action that was received.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    func workflowDidReceiveAction<Action: WorkflowAction>(
+        _ action: Action,
```

Comment:
can the workflow itself be passed in here, and its state?

Reply:
it probably can – i think i had experimented with some variants of that, and seem to recall some issues that weren't straightforward to resolve, but i will take a look at it again.